KAFKA-1595; Remove deprecated and slower Scala JSON parser

ijuma opened the pull request:
In a test by @onurkaraman involving 3066 topics and 95895 partitions,
Controller initialisation time spent on JSON parsing would be reduced from
37.1 seconds to 0.7 seconds by switching from the current JSON parser to
Jackson. See the following JIRA comment for more details:

https://issues.apache.org/jira/browse/KAFKA-5328?focusedCommentId=16027086

I tested that we only use Jackson methods introduced in 2.0 in the main
codebase by compiling it with the older version locally. We use a
constructor introduced in 2.4 in one test, but I didn't remove it as it
seemed harmless. The reasoning for this is explained in the mailing list
thread:

http://search-hadoop.com/m/uyzND1FWbWw1qUbWe

Finally, this PR only handles the parsing side. It would be good to use Jackson
for serialising to JSON as well. I filed KAFKA-5631 for that.

ijuma committed:
Introduce `testJsonParse`

Simple test that shows existing behaviour.

ijuma committed:
KAFKA-1595; Remove deprecated and slower scala JSON parser from kafka.consumer.TopicCount

A thin wrapper over Jackson's Tree Model API is used as the replacement. This wrapper
increases safety while providing a simple, but powerful API through the usage of the
`DecodeJson` type class. Even though this has a maintenance cost, it makes the API
much more convenient from Scala. A number of tests were added to verify the
behaviour of this wrapper.

The Scala module for Jackson doesn't provide any help for our current usage, so we don't
depend on it.

An attempt has been made to maintain the existing behaviour regarding when exceptions
are thrown. There are a number of cases where `JsonMappingException` will be thrown
instead of `ClassCastException`, however. It is expected that users would not try to catch
`ClassCastException`.

ijuma committed:
Minor clean-ups in `Json.encode`

ijuma commented:
Jenkins, retest this please

guozhangwang commented:
@ijuma Is this still relevant? If yes could you rebase?

ijuma commented:
@guozhangwang It is still relevant, but hopelessly out of date by now. :) I will close the PR until I have a chance to resubmit.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6037/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6022/
Test FAILed (JDK 8 and Scala 2.12).

ijuma committed:
Fix parsing of producer id JSON and clean-up the failing test

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6038/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6023/
Test PASSed (JDK 8 and Scala 2.12).

ijuma commented:
@onurkaraman I've rebased the branch. Would you like to review it?

ijuma commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -89,39 +90,33 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
-          "{\"version\" : 2, \"entity_path\":\"entity_type/entity_name\"}." +
-          " Received: " + json)
+      // There are no config overrides.
+      // Ignore non-json notifications because they can be from the deprecated TopicConfigManager
```
Will clarify this comment as the new placement makes it less clear.

ijuma committed:
Remove incorrect comment

ijuma committed:
Formatting fixes

ijuma commented:
@hachikuji, if you have time, maybe you can take a look to see what you think of the approach?

ijuma committed:
Use string interpolation in a few places

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6301/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6285/
Test FAILed (JDK 8 and Scala 2.12).

ijuma committed:
Update newly introduced `getBrokerIdFromLogDirEvent` to work with new implementation of Json.parseFull

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6306/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6290/
Test PASSed (JDK 8 and Scala 2.12).

onurkaraman commented:
My bad @ijuma. I thought I had left comments shortly after you pinged me on the PR.

Sort of a high-level comment: It seems like the approach of using implicits would work. I think people have mixed opinions on implicits in scala. Have you considered a more explicit approach? For instance, using jackson to convert to-and-from versioned POJOs. I think this would end up having clearly defined data formats for each znode path similar to our Protocol class for wire format definitions.

ijuma commented:
Thanks for the review @onurkaraman. There are many types of implicit usage in Scala. In this case I'm using implicits as type classes. This is generally considered as one of the most useful features in Scala (and Haskell). Implicit conversions are the ones that can be misused at times (they are still useful as extension methods).

I did consider versioned POJOs and I agree that it would be nice to have clearly defined formats. That's a bigger change, however and there's a higher potential for breaking compatibility with older formats depending on the quality of our tests (I'm a bit suspicious). We'd have to depend on the jackson-scala module if we use Scala classes in the POJOs, but we could potentially stick to Java classes.

 My thinking is that we should first convert to Jackson while using the same approach as we do now. It's easy to review and gives us the performance, concurrency and additional type safety benefit.

In a second phase, we should look into using versioned classes to represent the data types. At that point, we can also use Jackson to serialize to JSON. We can also do this incrementally as the underlying library is the same, we are just changing the implementation.

Thoughts?

onurkaraman commented:
@ijuma: Makes sense. I agree.

onurkaraman commented:
LGTM.

ijuma commented:
@rajinisivaram, maybe you can review this? It's mostly mechanical changes with some thin wrappers over Jackson.

ijuma committed:
Remove unused scalaParserCombinators from dependencies.gradle

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6348/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6333/
Test PASSed (JDK 8 and Scala 2.12).

rajinisivaram commented on core/src/main/scala/kafka/utils/Json.scala:
```diff
@@ -14,34 +14,27 @@
-  
+  def parseFull(input: String): Option[JsonValue] =
+    try Option(mapper.readTree(input)).map(JsonValue(_))
+    catch { case e: JsonProcessingException => None }
```
Isn't this swallowing an exception while the old code used to propagate parsing errors? Don't callers just ignore `None`? It looks like a deliberate change, but I wasn't sure why.

ijuma replied:
Thanks for the review @rajinisivaram. The behaviour is the same. The catch block in the old code doesn't seem to do anything. For example:

```scala
scala> JSON.parseFull("}")
res1: Option[Any] = None
```

ijuma replied:
The example above shows that `None` is returned for invalid JSON.

rajinisivaram replied:
@ijuma Thanks, LGTM.

rajinisivaram reviewed (approved):
Left a minor comment, but the changes look good.

ijuma committed:
Remove `JsonValue.toOption` since `to[Option[Foo]]` works

ijuma committed:
Add scaladoc

ijuma committed:
Handle null rack in broker json properly

ijuma committed:
Clean-up `BrokerEndPointTest`

ijuma commented:
@rajinisivaram, I noticed a bug and fixed it here (with tests): https://github.com/apache/kafka/pull/83/commits/30c9d02e0b0c427a91843f37f9ff9c91826157cd

I also added scaladoc to the parsing classes and removed the unnecessary `JsonValue.toOption` method.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6384/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6369/
Test PASSed (JDK 8 and Scala 2.12).

rajinisivaram reviewed (approved):
LGTM

ijuma commented:
Thanks for the reviews, merging to trunk.

leonhong commented:
great pr
